### PR TITLE
Avoid converting BTreeMap to Vec in TantivyDocument::add_object.

### PR DESCRIPTION
**Pull request by adamreichold**
For the next breaking release when the immediate issues with 0.22.x are resolved.

### CODE REVIEW

**Comment by fulmicoton** [comment]
@adamreichold it won't compile.

**Comment by adamreichold** [comment]
> @adamreichold it won't compile.

Yes, because a sequence of key-value pairs cannot directly be deserialized from a map. I added a thin wrapper to make that adjustment which avoids the overhead of the associative data structures, but keeps the convenience during deserialization.

File: src/schema/document/default_document.rs [hunk 1]
```diff
@@ -150,7 +151,7 @@ impl TantivyDocument {
     }
 
     /// Add a dynamic object field
-    pub fn add_object(&mut self, field: Field, object: BTreeMap<String, OwnedValue>) {
+    pub fn add_object(&mut self, field: Field, FlatObject(object): FlatObject) {
```
Comment (PSeitz) [inline src/schema/document/default_document.rs:123]: that seems like a worse API

Comment (adamreichold) [reply src/schema/document/default_document.rs:123]: Please expand on worse in what way or rather what you would prefer?

It definitely adds a bit of ceremony for the wrapper type, but it does avoid the unnecessary conversion for example in the unit tests here. Would you prefer some kind of implicit conversion via generics and trait bounds?

Comment (PSeitz) [reply src/schema/document/default_document.rs:123]: `BTreeMap` is a well know data structure for objects, while `FlatObject` would require some digging on how to convert into it

Comment (fulmicoton) [reply src/schema/document/default_document.rs:123]: what about any `impl IntoIterator<Item=(String, OwnedValue>)>`

**Comment by fulmicoton** [comment]
I don't quite understand the motivation for this PR. @adamreichold ?

**Comment by adamreichold** [comment]
> I don't quite understand the motivation for this PR. @adamreichold ?

`TantivyDocument` stores objects as flat vectors of key-value pairs but the API here requires going through `BTreeMap` only to convert it internally. To me, this seems wasteful and while I do not want to specifically avoid `BTreeMap`, I would like the API to enable at least one way of using it without the unnecessary intermediate collection.

**Review by PSeitz — COMMENTED** [review]
add_object is a convenience function, you can now add anything that implements `Value`

**Comment by adamreichold** [comment]
> add_object is a convenience function, you can now add anything that implements `Value`

Agreed. I was somewhat mechanically rebasing my branches and did not review all of the API changes implied by `CompactDoc`. Closing this one.

**Pull request closed without merging**